### PR TITLE
Update grpc-gataway (supporting response_body in google.api.HttpRule)

### PR DESCRIPTION
**Pull request by doroginin**
Also:
 * use AST during generation implementation
 * fix generation in case of snake_case in service name or request/response type

### CODE REVIEW

File: cmd/protoc-gen-goclay/internal/snake.go [hunk 1]
```diff
@@ -0,0 +1,23 @@
+package internal
+
+import (
+	"unicode"
+)
+
+// ToSnake convert the given string to snake case following the Golang format:
+// acronyms are converted to lower-case and preceded by an underscore.
+// copied form https://gist.github.com/elwinar/14e1e897fdbe4d3432e1
+func ToSnake(in string) string {
```
Comment (mwf) [inline cmd/protoc-gen-goclay/internal/snake.go]: Let's copy-paste from https://github.com/azer/snakecase

It's a slightly improved version of the same gist.

I've already benchmarked 4 different libs here - https://github.com/mwf/goplay/tree/master/snakecase 
And it's the best an the simplest at the same time :)

File: cmd/protoc-gen-goclay/genhandler/template.go [hunk 2]
```diff
@@ -80,8 +86,12 @@ func applyDescTemplate(p param) (string, error) {
 
 func goTypeName(s string) string {
```
Comment (mwf) [inline cmd/protoc-gen-goclay/genhandler/template.go]: Let's add some tests :)

File: cmd/protoc-gen-goclay/genhandler/handler.go [hunk 3]
```diff
@@ -309,6 +318,44 @@ func fileExists(path string) bool {
 	return false
 }
 
+func typeExists(typeName string, pkg *ast.Package) bool {
```
[the remark this answers is not in the document]
Comment (doroginin) [reply cmd/protoc-gen-goclay/genhandler/handler.go]: done

File: cmd/protoc-gen-goclay/genhandler/template.go (hunk 2 above)
Comment (doroginin) [reply cmd/protoc-gen-goclay/genhandler/template.go]: done

File: cmd/protoc-gen-goclay/genhandler/handler.go [hunk 4]
```diff
@@ -105,37 +108,44 @@ func (g *Generator) Generate(targets []*descriptor.File) ([]*plugin.CodeGenerato
 		glog.V(1).Infof("Will emit %s", output)
 
 		if g.options.Impl {
-			output := fmt.Sprintf(filepath.Join(goPkg, g.options.ImplPath, "%s.pb.impl.go"), base)
-			output = filepath.Clean(output)
+			fileSet := token.NewFileSet()
```
Comment (utrack) [inline cmd/protoc-gen-goclay/genhandler/handler.go]: не слишком жирно ли всё прятать в один for?

Comment (doroginin) [reply cmd/protoc-gen-goclay/genhandler/handler.go]: отрефакторил!

**Pull request merged**